From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mio <mio@mio19.uk>
Date: Thu, 22 Oct 2026 23:02:08 +1100
Subject: [PATCH] devices/fossil-hoki: update vendor kernel to 4.14.212

Keep each stable increment as a separate patch and advance through 4.14.210, 4.14.211, and 4.14.212 without concatenating patch text.
---
 devices/fossil-hoki/kernel/default.nix        |  85 +++++++++-
 .../kernel/linux-4.14.209-210.patch.patch     |  28 ++++
 .../kernel/linux-4.14.211-212.patch.patch     | 152 ++++++++++++++++++
 3 files changed, 264 insertions(+), 1 deletion(-)
 create mode 100644 devices/fossil-hoki/kernel/linux-4.14.209-210.patch.patch
 create mode 100644 devices/fossil-hoki/kernel/linux-4.14.211-212.patch.patch

diff --git a/devices/fossil-hoki/kernel/default.nix b/devices/fossil-hoki/kernel/default.nix
index 6f61e0b5..b87a910e 100644
--- a/devices/fossil-hoki/kernel/default.nix
+++ b/devices/fossil-hoki/kernel/default.nix
@@ -82,6 +82,84 @@ let
         cp "$TMPDIR/stable/linux-4.14.206-207.patch" "$out"
       '';
 
+  # Incremental 4.14.207 -> 4.14.208 stable patch.
+  stablePatch208 =
+    runCommand "linux-4.14.207-208.patch"
+      {
+        nativeBuildInputs = [ xz ];
+        src = fetchurl {
+          url = "https://cdn.kernel.org/pub/linux/kernel/v4.x/incr/patch-4.14.207-208.xz";
+          hash = "sha256-RUgB0ik3YaGix3k3Tm5VJbIMaSvZIp/+fuCNIuOguHk=";
+        };
+      }
+      ''
+        xz -dc "$src" > "$out"
+      '';
+
+  # Incremental 4.14.208 -> 4.14.209 stable patch.
+  stablePatch209 =
+    runCommand "linux-4.14.208-209.patch"
+      {
+        nativeBuildInputs = [ xz ];
+        src = fetchurl {
+          url = "https://cdn.kernel.org/pub/linux/kernel/v4.x/incr/patch-4.14.208-209.xz";
+          hash = "sha256-r0AtMyS92An0LizlWgoTbudht5/UD+f+MP38MlBQjQY=";
+        };
+      }
+      ''
+        xz -dc "$src" > "$out"
+      '';
+
+  # Incremental 4.14.209 -> 4.14.210 stable patch.
+  stablePatch210 =
+    runCommand "linux-4.14.209-210.patch"
+      {
+        nativeBuildInputs = [ xz ];
+        src = fetchurl {
+          url = "https://cdn.kernel.org/pub/linux/kernel/v4.x/incr/patch-4.14.209-210.xz";
+          hash = "sha256-w4lZp53kPfuo1Gtfs0xkB1Q5kRZ+VaNc7NfEGqKpVn4=";
+        };
+        patchFile = ./linux-4.14.209-210.patch.patch;
+      }
+      ''
+        mkdir -p "$TMPDIR/stable"
+        xz -dc "$src" > "$TMPDIR/stable/linux-4.14.209-210.patch"
+        (cd "$TMPDIR/stable" && patch -p1 -i "$patchFile")
+        cp "$TMPDIR/stable/linux-4.14.209-210.patch" "$out"
+      '';
+
+  # Incremental 4.14.210 -> 4.14.211 stable patch.
+  stablePatch211 =
+    runCommand "linux-4.14.210-211.patch"
+      {
+        nativeBuildInputs = [ xz ];
+        src = fetchurl {
+          url = "https://cdn.kernel.org/pub/linux/kernel/v4.x/incr/patch-4.14.210-211.xz";
+          hash = "sha256-Lra7YGHSls61bxwvS+fM3ySspqJRdVsaZMWCb7OXynI=";
+        };
+      }
+      ''
+        xz -dc "$src" > "$out"
+      '';
+
+  # Incremental 4.14.211 -> 4.14.212 stable patch.
+  stablePatch212 =
+    runCommand "linux-4.14.211-212.patch"
+      {
+        nativeBuildInputs = [ xz ];
+        src = fetchurl {
+          url = "https://cdn.kernel.org/pub/linux/kernel/v4.x/incr/patch-4.14.211-212.xz";
+          hash = "sha256-2rgzDbH88Z92a+1SIMwW6LhHS2TlWIwOW2w2AK6DWsw=";
+        };
+        patchFile = ./linux-4.14.211-212.patch.patch;
+      }
+      ''
+        mkdir -p "$TMPDIR/stable"
+        xz -dc "$src" > "$TMPDIR/stable/linux-4.14.211-212.patch"
+        (cd "$TMPDIR/stable" && patch -p1 -i "$patchFile")
+        cp "$TMPDIR/stable/linux-4.14.211-212.patch" "$out"
+      '';
+
   # Patch upstream defconfig to enable required Bluetooth protocols (RFCOMM, BNEP, HIDP),
   # Btrfs filesystem support, and transparent Zstandard/XXHash compression.
   configfile =
@@ -96,7 +174,7 @@ let
 in
 
 kernel-builder-gcc8 {
-  version = "4.14.207";
+  version = "4.14.212";
   inherit configfile;
 
   src = fetchFromGitHub {
@@ -108,6 +186,11 @@ kernel-builder-gcc8 {
 
   patches = [
     stablePatch
+    stablePatch208
+    stablePatch209
+    stablePatch210
+    stablePatch211
+    stablePatch212
     (upstreamPatch "0001-dts-Add-hoki-device-trees.patch" "0z3c6lbqpqmhqsy158wbpl0zcfkr94nwh42d6px309684qq4c8v2")
     (upstreamPatch "0002-mmc-Fix-embedded_sdio_data-duplicate-definition.patch" "0h7d3pxz8di2rh257kmc83iwcpf8xmrqhbqq4an2vclfdhyr297j")
     (upstreamPatch "0003-video-fbdev-msm-Provide-mdss_dsi_switch_page.patch" "1igqg795cvr81w948k4dj8nrbzc2pz5bg6gdp68jfplm2hcs4nch")
diff --git a/devices/fossil-hoki/kernel/linux-4.14.209-210.patch.patch b/devices/fossil-hoki/kernel/linux-4.14.209-210.patch.patch
new file mode 100644
index 00000000..f7be7ed5
--- /dev/null
+++ b/devices/fossil-hoki/kernel/linux-4.14.209-210.patch.patch
@@ -0,0 +1,28 @@
+# Drop the ufshcd shutdown hunk from 4.14.209 -> 4.14.210;
+# the vendor tree already uses pm_runtime_get_sync().
+--- a/linux-4.14.209-210.patch
++++ b/linux-4.14.209-210.patch
+@@ -1229,23 +1229,6 @@
+  		task->have_checked_conn = true;
+  		rc = BLK_EH_RESET_TIMER;
+  		goto done;
+-diff --git a/drivers/scsi/ufs/ufshcd.c b/drivers/scsi/ufs/ufshcd.c
+-index c1792f271ac5d..a3a3ee6e2a002 100644
+---- a/drivers/scsi/ufs/ufshcd.c
+-+++ b/drivers/scsi/ufs/ufshcd.c
+-@@ -7792,11 +7792,7 @@ int ufshcd_shutdown(struct ufs_hba *hba)
+- 	if (ufshcd_is_ufs_dev_poweroff(hba) && ufshcd_is_link_off(hba))
+- 		goto out;
+- 
+--	if (pm_runtime_suspended(hba->dev)) {
+--		ret = ufshcd_runtime_resume(hba);
+--		if (ret)
+--			goto out;
+--	}
+-+	pm_runtime_get_sync(hba->dev);
+- 
+- 	ret = ufshcd_suspend(hba, UFS_SHUTDOWN_PM);
+- out:
+ diff --git a/drivers/target/iscsi/iscsi_target.c b/drivers/target/iscsi/iscsi_target.c
+ index da80c03de6ea4..d9fcef82ddf59 100644
+ --- a/drivers/target/iscsi/iscsi_target.c
diff --git a/devices/fossil-hoki/kernel/linux-4.14.211-212.patch.patch b/devices/fossil-hoki/kernel/linux-4.14.211-212.patch.patch
new file mode 100644
index 00000000..6e5dee43
--- /dev/null
+++ b/devices/fossil-hoki/kernel/linux-4.14.211-212.patch.patch
@@ -0,0 +1,152 @@
+# Drop the tty_jobctrl and f_fs diffs from 4.14.211 -> 4.14.212;
+# the vendor tree already has the f_fs change and the tty hunk does not apply.
+--- a/linux-4.14.211-212.patch
++++ b/linux-4.14.211-212.patch
+@@ -621,147 +621,6 @@
+  #endif
+  }
+  
+-diff --git a/drivers/tty/tty_jobctrl.c b/drivers/tty/tty_jobctrl.c
+-index e7032309ee871..b71e61e79c5a3 100644
+---- a/drivers/tty/tty_jobctrl.c
+-+++ b/drivers/tty/tty_jobctrl.c
+-@@ -102,8 +102,8 @@ static void __proc_set_tty(struct tty_struct *tty)
+- 	put_pid(tty->session);
+- 	put_pid(tty->pgrp);
+- 	tty->pgrp = get_pid(task_pgrp(current));
+--	spin_unlock_irqrestore(&tty->ctrl_lock, flags);
+- 	tty->session = get_pid(task_session(current));
+-+	spin_unlock_irqrestore(&tty->ctrl_lock, flags);
+- 	if (current->signal->tty) {
+- 		tty_debug(tty, "current tty %s not NULL!!\n",
+- 			  current->signal->tty->name);
+-@@ -292,20 +292,23 @@ void disassociate_ctty(int on_exit)
+- 	spin_lock_irq(&current->sighand->siglock);
+- 	put_pid(current->signal->tty_old_pgrp);
+- 	current->signal->tty_old_pgrp = NULL;
+--
+- 	tty = tty_kref_get(current->signal->tty);
+-+	spin_unlock_irq(&current->sighand->siglock);
+-+
+- 	if (tty) {
+- 		unsigned long flags;
+-+
+-+		tty_lock(tty);
+- 		spin_lock_irqsave(&tty->ctrl_lock, flags);
+- 		put_pid(tty->session);
+- 		put_pid(tty->pgrp);
+- 		tty->session = NULL;
+- 		tty->pgrp = NULL;
+- 		spin_unlock_irqrestore(&tty->ctrl_lock, flags);
+-+		tty_unlock(tty);
+- 		tty_kref_put(tty);
+- 	}
+- 
+--	spin_unlock_irq(&current->sighand->siglock);
+- 	/* Now clear signal->tty under the lock */
+- 	read_lock(&tasklist_lock);
+- 	session_clear_tty(task_session(current));
+-@@ -476,14 +479,19 @@ static int tiocspgrp(struct tty_struct *tty, struct tty_struct *real_tty, pid_t
+- 		return -ENOTTY;
+- 	if (retval)
+- 		return retval;
+--	if (!current->signal->tty ||
+--	    (current->signal->tty != real_tty) ||
+--	    (real_tty->session != task_session(current)))
+--		return -ENOTTY;
+-+
+- 	if (get_user(pgrp_nr, p))
+- 		return -EFAULT;
+- 	if (pgrp_nr < 0)
+- 		return -EINVAL;
+-+
+-+	spin_lock_irq(&real_tty->ctrl_lock);
+-+	if (!current->signal->tty ||
+-+	    (current->signal->tty != real_tty) ||
+-+	    (real_tty->session != task_session(current))) {
+-+		retval = -ENOTTY;
+-+		goto out_unlock_ctrl;
+-+	}
+- 	rcu_read_lock();
+- 	pgrp = find_vpid(pgrp_nr);
+- 	retval = -ESRCH;
+-@@ -493,12 +501,12 @@ static int tiocspgrp(struct tty_struct *tty, struct tty_struct *real_tty, pid_t
+- 	if (session_of_pgrp(pgrp) != task_session(current))
+- 		goto out_unlock;
+- 	retval = 0;
+--	spin_lock_irq(&tty->ctrl_lock);
+- 	put_pid(real_tty->pgrp);
+- 	real_tty->pgrp = get_pid(pgrp);
+--	spin_unlock_irq(&tty->ctrl_lock);
+- out_unlock:
+- 	rcu_read_unlock();
+-+out_unlock_ctrl:
+-+	spin_unlock_irq(&real_tty->ctrl_lock);
+- 	return retval;
+- }
+- 
+-@@ -510,20 +518,30 @@ out_unlock:
+-  *
+-  *	Obtain the session id of the tty. If there is no session
+-  *	return an error.
+-- *
+-- *	Locking: none. Reference to current->signal->tty is safe.
+-  */
+- static int tiocgsid(struct tty_struct *tty, struct tty_struct *real_tty, pid_t __user *p)
+- {
+-+	unsigned long flags;
+-+	pid_t sid;
+-+
+- 	/*
+- 	 * (tty == real_tty) is a cheap way of
+- 	 * testing if the tty is NOT a master pty.
+- 	*/
+- 	if (tty == real_tty && current->signal->tty != real_tty)
+- 		return -ENOTTY;
+-+
+-+	spin_lock_irqsave(&real_tty->ctrl_lock, flags);
+- 	if (!real_tty->session)
+--		return -ENOTTY;
+--	return put_user(pid_vnr(real_tty->session), p);
+-+		goto err;
+-+	sid = pid_vnr(real_tty->session);
+-+	spin_unlock_irqrestore(&real_tty->ctrl_lock, flags);
+-+
+-+	return put_user(sid, p);
+-+
+-+err:
+-+	spin_unlock_irqrestore(&real_tty->ctrl_lock, flags);
+-+	return -ENOTTY;
+- }
+- 
+- /*
+-diff --git a/drivers/usb/gadget/function/f_fs.c b/drivers/usb/gadget/function/f_fs.c
+-index 2ff7c21bbda58..2f98161d062e2 100644
+---- a/drivers/usb/gadget/function/f_fs.c
+-+++ b/drivers/usb/gadget/function/f_fs.c
+-@@ -1244,7 +1244,7 @@ static long ffs_epfile_ioctl(struct file *file, unsigned code,
+- 	case FUNCTIONFS_ENDPOINT_DESC:
+- 	{
+- 		int desc_idx;
+--		struct usb_endpoint_descriptor *desc;
+-+		struct usb_endpoint_descriptor desc1, *desc;
+- 
+- 		switch (epfile->ffs->gadget->speed) {
+- 		case USB_SPEED_SUPER:
+-@@ -1256,10 +1256,12 @@ static long ffs_epfile_ioctl(struct file *file, unsigned code,
+- 		default:
+- 			desc_idx = 0;
+- 		}
+-+
+- 		desc = epfile->ep->descs[desc_idx];
+-+		memcpy(&desc1, desc, desc->bLength);
+- 
+- 		spin_unlock_irq(&epfile->ffs->eps_lock);
+--		ret = copy_to_user((void *)value, desc, desc->bLength);
+-+		ret = copy_to_user((void *)value, &desc1, desc1.bLength);
+- 		if (ret)
+- 			ret = -EFAULT;
+- 		return ret;
+ diff --git a/drivers/usb/serial/ch341.c b/drivers/usb/serial/ch341.c
+ index d45a3f4e752c7..389b24b84ba46 100644
+ --- a/drivers/usb/serial/ch341.c
